pkg/pillar/zedcloud/deferred.go

```diff
@@ -24,132 +27,164 @@ import (
+	itemType      interface{}
+	key           string
 	buf           *bytes.Buffer
 	size          int64
 	url           string
-	zedcloudCtx   *ZedCloudContext
 	bailOnHTTPErr bool // Return 4xx and 5xx without trying other interfaces
 }
 
-type deferredItemList struct {
-	list []deferredItem
-}
-
+const maxTimeToHandleDeferred = time.Minute
 const longTime1 = time.Hour * 24
 const longTime2 = time.Hour * 48
 
 // DeferredContext is part of ZedcloudContext
 type DeferredContext struct {
-	deferredItems map[string]deferredItemList
-	ticker        flextimer.FlexTickerHandle
+	deferredItems          []*deferredItem
+	ticker                 flextimer.FlexTickerHandle
+	priorityCheckFunctions []TypePriorityCheckFunction
+	lock                   *sync.Mutex
+	sentHandler            *SentHandlerFunction
+	zedcloudCtx            *ZedCloudContext
+	iteration              int
 }
 
+//TypePriorityCheckFunction returns true in case of find type with high priority
+type TypePriorityCheckFunction func(itemType interface{}) bool
+
+//SentHandlerFunction allow doing something with data if it was handled
+//result indicates sending result
+type SentHandlerFunction func(itemType interface{}, data *bytes.Buffer, result types.SenderResult)
+
 // GetDeferredChan creates and returns a channel to the caller
 // We always keep a flextimer running so that we can return
 // the associated channel. We adjust the times when we start and stop
 // the timer.
-func GetDeferredChan(zedcloudCtx *ZedCloudContext) <-chan time.Time {
+// sentHandler is callback which will be run on successful sent
+// priorityCheckFunctions may be added to send item with matched itemType firstly
+// default function at the end of priorityCheckFunctions added to serve non-priority items
+func GetDeferredChan(zedcloudCtx *ZedCloudContext, sentHandler *SentHandlerFunction, priorityCheckFunctions ...TypePriorityCheckFunction) <-chan time.Time {
+	//append with return first
+	priorityCheckFunctions = append(priorityCheckFunctions, func(obj interface{}) bool {
+		return true
+	})
 	zedcloudCtx.deferredCtx = DeferredContext{
-		deferredItems: make(map[string]deferredItemList),
-		ticker:        flextimer.NewRangeTicker(longTime1, longTime2),
+		lock:                   &sync.Mutex{},
+		ticker:                 flextimer.NewRangeTicker(longTime1, longTime2),
+		sentHandler:            sentHandler,
+		priorityCheckFunctions: priorityCheckFunctions,
+		zedcloudCtx:            zedcloudCtx,
 	}
 	return zedcloudCtx.deferredCtx.ticker.C
 }
 
-// Try to send all deferred items. Give up if any one fails
+// HandleDeferred try to send all deferred items (or only one if sendOne set). Give up if any one fails
 // Stop timer if map becomes empty
 // Returns true when there are no more deferred items
-func HandleDeferred(zedcloudCtx *ZedCloudContext, event time.Time, spacing time.Duration) bool {
+func HandleDeferred(zedcloudCtx *ZedCloudContext, event time.Time, spacing time.Duration, sendOne bool) bool {
```

Comment:
One cleanup item: the event argument isn't really needed.
Instead handleDeferred can just set start := time.Now() before its loop.